File: AlerteTCL/Models/Vehicle.swift
import Foundation
import CoreLocation
import SwiftUI

struct Vehicle: Identifiable, Hashable {
    let id: String
    let latitude: Double
    let longitude: Double
    let bearing: Double
    let lineRef: String
    let lineName: String
    let vehicleType: VehicleType
    let destination: String
    /// Sens SIRI normalisé : "A" aller, "R" retour, "" inconnu (cf. DirectionMatching).
    let direction: String
    let delay: Int
    let status: String?
    let recordedAt: Date?
    let validUntil: Date?
    let nextStop: StopInfo?
    
    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }
    
    var delayFormatted: String {
        if delay == 0 {
            return "À l'heure"
        } else if delay > 0 {
            let minutes = delay / 60
            if minutes > 0 {
                return "+\(minutes) min"
            } else {
                return "+\(delay) sec"
            }
        } else {
            let minutes = abs(delay) / 60
            if minutes > 0 {
                return "-\(minutes) min"
            } else {
                return "\(delay) sec"
            }
        }
    }
    
    var isDelayed: Bool {
        delay > 60
    }
    
    var isEarly: Bool {
        delay < -60
    }

    /// Numéro de parc extrait du VehicleRef SIRI (ex. "ActIV:Vehicle:Bus:1512:LOC" → "1512").
    var fleetNumber: String? {
        let parts = id.split(separator: ":")
        guard parts.count > 3 else { return nil }
        return String(parts[3])
    }

    /// Âge de la dernière position transmise par TCL (RecordedAtTime SIRI), en secondes.
    var positionAge: TimeInterval? {
        recordedAt.map { max(0, Date().timeIntervalSince($0)) }
    }

    /// Fraîcheur de la position, pour l'affichage (couleur + transparence).
    /// Seuils : le flux SIRI TCL republie chaque véhicule toutes les ~15-60 s ;
    /// au-delà de 2 min la position est considérée obsolète (véhicule "fantôme").
    enum PositionFreshness {
        case fresh   // < 45 s
        case aging   // 45 s – 2 min
        case stale   // > 2 min
    }

    var positionFreshness: PositionFreshness {
        guard let age = positionAge else { return .fresh }
        if age < 45 { return .fresh }
        if age < 120 { return .aging }
        return .stale
    }

    /// Formate un âge en texte court ("12 s", "1 min 30", "4 min").
    static func formattedAge(_ seconds: TimeInterval) -> String {
        let s = Int(seconds)
        if s < 60 { return "\(s) s" }
        let m = s / 60
        let r = s % 60
        if m >= 5 || r == 0 { return "\(m) min" }
        return "\(m) min \(String(format: "%02d", r))"
    }
}

struct StopInfo: Identifiable, Hashable {
    let id: String
    let stopRef: String
    let stopName: String?
    let aimedArrivalTime: Date?
    let aimedDepartureTime: Date?
    let distanceFromStop: Int?
    let order: Int?
    
    var timeUntilArrival: TimeInterval? {
        guard let arrivalTime = aimedArrivalTime else { return nil }
        return arrivalTime.timeIntervalSinceNow
    }
    
}

enum VehicleType: String, CaseIterable {
    case tram = "Tramway"
    case bus = "Bus"
    case trolley = "Trolleybus"
    case funicular = "Funiculaire"
    case metro = "Métro"
    case navigone = "Navigone"
    
    var icon: String {
        switch self {
        case .tram: return "tram"
        case .bus: return "bus.fill"
        case .trolley: return "bus.doubledecker.fill"
        case .funicular: return "cablecar.fill"
        case .metro: return "tram.fill"
        case .navigone: return "ferry.fill"
        }
    }
    
    var sortOrder: Int {
        switch self {
        case .metro: return 0
        case .funicular: return 1
        case .tram: return 2
        case .trolley: return 3
        case .bus: return 4
        case .navigone: return 5
        }
    }
}

struct SIRIResponse: Codable {
    let Siri: SIRIRoot
}

struct SIRIRoot: Codable {
    let ServiceDelivery: ServiceDelivery
}

struct ServiceDelivery: Codable {
    let ResponseTimestamp: String?
    let VehicleMonitoringDelivery: [VehicleMonitoringDelivery]?
    /// `true` = réponse tronquée, d'autres véhicules sont disponibles.
    /// Ne devrait pas arriver avec MaximumVehicles=2000 dans le worker.
    let MoreData: Bool?
}

struct VehicleMonitoringDelivery: Codable {
    let ResponseTimestamp: String?
    let VehicleActivity: [VehicleActivity]?
}

struct VehicleActivity: Codable {
    let RecordedAtTime: String?
    let ValidUntilTime: String?
    let VehicleMonitoringRef: RefValue?
    let MonitoredVehicleJourney: MonitoredVehicleJourney?
}

struct RefValue: Codable {
    let value: String?
}

struct MonitoredVehicleJourney: Codable {
    let LineRef: RefValue?
    let DirectionRef: RefValue?
    let DestinationRef: RefValue?
    let VehicleRef: RefValue?
    let VehicleLocation: VehicleLocation?
    let Bearing: Double?
    let Delay: String?
    let VehicleStatus: String?
    let MonitoredCall: MonitoredCall?
}

struct MonitoredCall: Codable {
    let AimedArrivalTime: String?
    let ActualArrivalTime: String?
    let ArrivalStatus: String?
    let AimedDepartureTime: String?
    let ActualDepartureTime: String?
    let DepartureStatus: String?
    let DistanceFromStop: Int?
    let StopPointRef: RefValue?
    let Order: Int?
}

struct VehicleLocation: Codable {
    let Longitude: Double?
    let Latitude: Double?
}

File: shared/src/commonMain/kotlin/com/alertetcl/shared/models/Vehicle.kt
package com.alertetcl.shared.models

import com.alertetcl.shared.geo.LatLng
import kotlinx.serialization.Serializable

@Serializable
enum class VehicleType(val displayName: String, val iconKey: String, val sortOrder: Int) {
    METRO("Métro", "metro", 0),
    FUNICULAR("Funiculaire", "funicular", 1),
    TRAM("Tramway", "tram", 2),
    TROLLEY("Trolleybus", "trolley", 3),
    NAVIGONE("Navigo'ne", "navigone", 4),
    BUS("Bus", "bus", 5);
}

@Serializable
data class StopInfo(
    val id: String,
    val stopRef: String,
    val stopName: String? = null,
    val aimedArrivalTimeEpoch: Long? = null,
    val aimedDepartureTimeEpoch: Long? = null,
    val distanceFromStop: Int? = null,
    val order: Int? = null
) {
    /** Identifiant TCL de l'arrêt (« ActIV:StopArea:SP:11518:SYTRAL » → 11518), commun au GeoServer et aux fiches horaires. */
    val numericId: Int? get() = stopRef.split(":").getOrNull(3)?.toIntOrNull() ?: stopRef.toIntOrNull()
}

@Serializable
data class Vehicle(
    val id: String,
    val latitude: Double,
    val longitude: Double,
    val bearing: Double,
    val lineRef: String,
    val lineName: String,
    val vehicleType: VehicleType,
    val destination: String,
    /** Sens SIRI normalisé : "A" aller, "R" retour, "" inconnu (cf. DirectionMatching). */
    val direction: String = "",
    val delay: Int,
    val status: String? = null,
    val recordedAtEpoch: Long? = null,
    val validUntilEpoch: Long? = null,
    val nextStop: StopInfo? = null
) {
    val coordinate: LatLng get() = LatLng(latitude, longitude)

    /** Numéro de parc extrait du VehicleRef SIRI (ex. "ActIV:Vehicle:Bus:1512:LOC" → "1512"). */
    val fleetNumber: String? get() = id.split(":").getOrNull(3)?.takeIf { it.isNotEmpty() }

    /** Ponctualité : le chiffre, ce qu'il veut dire, et la phrase entière (cf. [VehicleTexts]). */
    val delayAmount: String get() = VehicleTexts.punctualityAmount(delay)
    val delayCaption: String get() = VehicleTexts.punctualityCaption(delay)
    val delayText: String get() = VehicleTexts.punctuality(delay)

    val isDelayed: Boolean get() = VehicleTexts.isDelayed(delay)
    val isEarly:   Boolean get() = VehicleTexts.isEarly(delay)

    /**
     * Arrivée au prochain arrêt (epoch, secondes) : son horaire prévu corrigé du retard constaté,
     * comme l'estimation de « Où est mon bus ». Null quand TCL ne donne pas d'horaire pour cet arrêt.
     */
    val nextStopArrivalEpoch: Long? get() = nextStop?.let { stop ->
        (stop.aimedArrivalTimeEpoch ?: stop.aimedDepartureTimeEpoch)?.plus(delay)
    }

    /** Légende de cette heure : « arrivée à Bellecour ». Null sans nom d'arrêt. */
    val nextStopArrivalCaption: String? get() = nextStop?.stopName?.let { VehicleTexts.arrivalCaption(it) }

    /** Âge de la dernière position transmise par TCL (RecordedAtTime SIRI), en secondes. */
    fun positionAgeSeconds(nowEpochMs: Long): Long? =
        recordedAtEpoch?.let { (nowEpochMs / 1000 - it).coerceAtLeast(0) }

    /**
     * Fraîcheur de la position, pour l'affichage (couleur de l'étiquette et de la fiche).
     * Le flux SIRI TCL republie chaque véhicule toutes les ~15-60 s ; passé
     * [HIDE_AFTER_SECONDS] sans nouvelle position, elle est obsolète : le véhicule
     * quitte la carte et sa fiche, si elle est ouverte, le signale.
     */
    fun positionFreshness(nowEpochMs: Long): PositionFreshness {
        val age = positionAgeSeconds(nowEpochMs) ?: return PositionFreshness.FRESH
        return when {
            age < 45                 -> PositionFreshness.FRESH
            age < HIDE_AFTER_SECONDS -> PositionFreshness.AGING
            else                     -> PositionFreshness.STALE
        }
    }

    /** False dès que la position est obsolète : le véhicule n'est plus dessiné sur la carte. */
    fun isShownOnMap(nowEpochMs: Long): Boolean = positionFreshness(nowEpochMs) != PositionFreshness.STALE

    companion object {
        /** Délai sans nouvelle position transmise par TCL au-delà duquel un véhicule disparaît de la carte. */
        const val HIDE_AFTER_SECONDS = 90L

        /** Formate un âge en texte court ("12 s", "1 min 30", "4 min"). */
        fun formattedAge(seconds: Long): String {
            if (seconds < 60) return "$seconds s"
            val m = seconds / 60
            val r = seconds % 60
            if (m >= 5 || r == 0L) return "$m min"
            return "$m min ${r.toString().padStart(2, '0')}"
        }
    }
}

enum class PositionFreshness {
    FRESH, AGING, STALE;

    /** Couleur d'état associée (jetons partagés, variantes claire et sombre). */
    val color: com.alertetcl.shared.design.ThemedColor get() = when (this) {
        FRESH -> com.alertetcl.shared.design.AppColors.fresh
        AGING -> com.alertetcl.shared.design.AppColors.aging
        STALE -> com.alertetcl.shared.design.AppColors.stale
    }
}
